executor/src/witgen/machines/machine_extractor.rs

```diff
@@ -197,15 +210,106 @@ pub fn split_out_machines<'a, T: FieldElement>(
+                        // If we already have witnesses on the LHS, include the LHS,
+                        // and vice-versa, but not across the "sides".
+                        let in_lhs = &self.refs_in_selected_expressions(left) & all_witnesses;
+                        let in_rhs = &self
+                            .refs_in_connection_rhs(&Connection::try_from(*i).unwrap())
+                            & all_witnesses;
+                        if in_lhs.intersection(&witnesses).next().is_some() {
+                            witnesses.extend(in_lhs);
+                        } else if in_rhs.intersection(&witnesses).next().is_some() {
+                            witnesses.extend(in_rhs);
+                        }
+                    }
+                    Identity::Connect(..) => {
+                        unimplemented!()
+                    }
+                };
+            }
+            if witnesses.len() == count {
+                return witnesses;
+            }
+        }
+    }
+
+    /// Like refs_in_selected_expressions(connection.right), but also includes the multiplicity column.
+    fn refs_in_connection_rhs(&self, connection: &Connection<T>) -> HashSet<PolyID> {
+        self.refs_in_selected_expressions(connection.right)
+            .into_iter()
+            .chain(connection.multiplicity_column)
+            .collect()
+    }
+
+    /// Extracts all references to names from selected expressions.
+    fn refs_in_selected_expressions(&self, sel_expr: &SelectedExpressions<T>) -> HashSet<PolyID> {
+        sel_expr
+            .children()
+            .flat_map(|e| self.refs_in_expression(e))
+            .collect()
+    }
+
+    /// Extracts all references to names from the "left" side of an identity. This is the left selected expressions for connecting identities, and everything for other identities.
+    fn refs_in_identity_left(&self, identity: &Identity<T>) -> HashSet<PolyID> {
+        match identity {
+            Identity::Lookup(LookupIdentity { left, .. })
+            | Identity::PhantomLookup(PhantomLookupIdentity { left, .. })
+            | Identity::Permutation(PermutationIdentity { left, .. })
+            | Identity::PhantomPermutation(PhantomPermutationIdentity { left, .. }) => {
+                self.refs_in_selected_expressions(left)
+            }
+            Identity::Polynomial(i) => self.refs_in_expression(&i.expression).collect(),
+            Identity::Connect(i) => i
+                .left
+                .iter()
+                .chain(&i.right)
+                .flat_map(|e| self.refs_in_expression(e))
+                .collect(),
+        }
+    }
+
+    fn refs_in_expression(&self, expr: &'a Expression<T>) -> impl Iterator<Item = PolyID> + '_ {
+        Box::new(expr.all_children().filter_map(move |e| match e {
+            Expression::Reference(p) => Some(p.poly_id),
+            _ => None,
+        }))
     }
```

Comment:
In #2007, we'll access `self.fixed` here to get the definitions of intermediate polynomials. Because this is a method of `MachineExtractor`, we can avoid passing it all the way down to his function.